preprocess_recovered_hormuud_messages.py

```diff
@@ -0,0 +1,270 @@
+
+import pytz
+from core_data_modules.logging import Logger
+from dateutil.parser import isoparse
+from rapid_pro_tools.rapid_pro_client import RapidProClient
+from storage.google_cloud import google_cloud_utils
+
+log = Logger(__name__)
+
+TARGET_SHORTCODE = "378"
+
+
+def get_incoming_hormuud_messages_from_rapid_pro(google_cloud_credentials_file_path, rapid_pro_domain,
+                                                 rapid_pro_token_file_url,
+                                                 created_after_inclusive=None, created_before_exclusive=None):
+    log.info("Downloading Rapid Pro access token...")
+    rapid_pro_token = google_cloud_utils.download_blob_to_string(
+        google_cloud_credentials_file_path, rapid_pro_token_file_url).strip()
+
+    rapid_pro = RapidProClient(rapid_pro_domain, rapid_pro_token)
+
+    all_messages = rapid_pro.get_raw_messages(
+        created_after_inclusive=created_after_inclusive,
+        created_before_exclusive=created_before_exclusive,
+        ignore_archives=True
+    )
+    log.info(f"Downloaded {len(all_messages)} messages")
+
+    log.info(f"Filtering for messages from URNs on Hormuud's networks")
+    hormuud_messages = [msg for msg in all_messages if msg.urn.startswith("tel:+25261") or msg.urn.startswith("tel:+25268")]
+    log.info(f"Filtered for messages from URNs on Hormuud's networks: {len(hormuud_messages)} messages remain")
+
+    log.info(f"Filtering for incoming messages")
+    incoming_hormuud_messages = [msg for msg in hormuud_messages if msg.direction == "in"]
+    log.info(f"Filtered for incoming messages: {len(incoming_hormuud_messages)} remain")
+
+    return incoming_hormuud_messages
+
+
+def get_incoming_hormuud_messages_from_recovery_csv(csv_path,
+                                                    received_after_inclusive=None, received_before_exclusive=None):
+    log.info(f"Loading recovered messages from Hormuud csv at {csv_path}...")
+    all_recovered_messages = []
+    with open(csv_path) as f:
+        reader = csv.DictReader(f)
+        for line in reader:
+            all_recovered_messages.append(line)
+    log.info(f"Loaded {len(all_recovered_messages)} messages")
+
+    log.info(f"Filtering for messages sent to the target short code {TARGET_SHORTCODE}...")
+    incoming_recovered_messages = [msg for msg in all_recovered_messages if msg["Receiver"] == TARGET_SHORTCODE]
+    log.info(f"Filtered for messages sent to the target short code {TARGET_SHORTCODE}: "
+             f"{len(incoming_recovered_messages)} recovered messages remain")
+
+    log.info(f"Standardising fieldnames")
+    for msg in incoming_recovered_messages:
+        msg["Sender"] = "tel:+" + msg["Sender"]
+        # Convert times with a try/catch because there are two possible formats due to the omission of ms when ms == 000
+        try:
+            msg["ReceivedOn"] = pytz.timezone("Africa/Mogadishu").localize(
+                datetime.strptime(msg["ReceivedOn"], "%d/%m/%Y %H:%M:%S.%f")
```

Comment:
This mangles the timestamps in place. You can fix using the version at https://github.com/AfricasVoices/Pool-CSAP-Somalia/pull/11/files, sorry.

Reply:
Updated, thanks.